rfcs/007-background-workers.md

```diff
@@ -0,0 +1,299 @@
+indicate that they are "alive" and ready to process jobs.
+
+`Versionstamp`: a 12 byte, unique, monotonically (but not sequentially)
+increasing value for each committed transaction.
+
+---
+
+# Detailed Description
+
+## Data Model
+
+The main job table:
+ - `("couch_workers", "jobs", JobType, JobId) = (JobState, WorkerId, Priority, CancelReq, HealthTimeout, JobInfo, JobOps)`
+
+Pending queue:
+ - `("couch_workers", "pending", JobType, Priority, JobId) = ""`
+
+Active jobs:
+ - `("couch_workers", "active", WorkerType, Worker, JobId) = (JobState, JobInfo, JobOpts)`
+
+Worker registration and health:
+ - `("couch_workers", WorkerType, "workers_vs") = Versionstamp`
+ - `("couch_workers", WorkerType, "workers", WorkerId) = WorkerOpts`
+ - `("couch_workers", WorkerType, "health", WorkerId) = HealthCounter`
+
+### Data model fields
+
+- `JobType`: The job type. It MAY be ``replication`` or ``indexing``, however,
+  specific types are not in scope of this document.
+
+- `JobId` : MUST uniquely identify a job in a cluster. It MAY be a `UUID` or
+  derived from some other job parameters. In case the same job is submitted
+  multiple times, it is up to the job creation and type-specific worker logic
+  to cancel, re-start or update jobs.
+
+- `JobState` : MUST be one of `pending`, `running`, or `completed`
+
+- `Priority` : MAY be any key which allows sorting jobs, such as a timestamp,
+  or a tag.
+
+- `WorkerId` : MUST uniquely identify a worker in a cluster. It MAY be
+  generated each time the worker is restarted, or it could be persisted across
+  worker restarts, as long as it is unique across the whole cluster.
+
+- `CancelReq` : A boolean flag indicating a request to the worker to cancel the
+  job.
+
+- `JobOps` : Type-specific job options. Represented as a JSON object. This MAY
+  contain fields like `"source"`, `"target"`, `"dbname"` etc.
+
+- `JobInfo` : A per-job info object. Represented as JSON. This object will
+  contain details about a job's current state. It MAY have fields such as
+  `"update_seq"`, `"error_count"`, etc.
+
+- `HealthTimeout` : Worker timeout defined how often the worker will update
+  they health entry to maintain liveness. The worker MUST update their health
+  entry at least this often to maintain liveness.
+
+- `HealthCounter` : Workers MUST updated this at least as often as the
+  `HealthTimeout` indicates. For efficiency, workers MAY use FDB atomic
+  increments here.
+
+
+## Job Lifecyle
+
+New jobs are posted to the main jobs table along with an entry in the pending
+queue. `WorkerId` is set to `nil`. `Priority` could be a coarse timestamp, for
+example "run this job at any time during the next 5 minutes", or it MAY be a
+compound key that has a timestamp + tag such that say continuous replications
+always sort before the normal ones. The uniqueness of `JobId` should reduce
+contention during insertion into the queue.
+
+Workers monitor the pending queue of their matching `JobType`. That is,
```

Comment:
reference or example of how workers monitor their pending queue would be nice ✌️

Reply:
There is now an API description that hopefully make it more clear. See the an accept() function in the worker API and its implementation.